File: TypeScript/assignments/Extra/Rpn.ts
export default function rpn(input: string) {
    const ops = input.split(' ');
    const result = ops.reduce(innerRpn, []);
    const finalValue = result.pop();
    return result.length === 0 ? finalValue : null;
}

function innerRpn(acc: number[], op: string): number[] {
    switch (op) {
        case "+":
        case "-":
        case "*":
        case "/":
        case "^":
            {
                const n1 = acc.pop() ?? 0.0;
                const n2 = acc.pop() ?? 0.0;
                switch (op) {
                    case "+":
                        acc.push(n2 + n1);
                        break;
                    case "-":
                        acc.push(n2 - n1);
                        break;
                    case "*":
                        acc.push(n2 * n1);
                        break;
                    case "/":
                        acc.push(n2 / n1);
                        break;
                    case "^":
                        acc.push(Math.pow(n2, n1));
                        break;
                }
            }
            break;
        case "ln":
            {
                const n1 = acc.pop() ?? 0.0;
                acc.push(Math.log(n1));
            }
            break;
        case "log10":
            {
                const n1 = acc.pop() ?? 0.0;
                acc.push(Math.log10(n1));
            }
            break;
        case "sum":
            {
                const sum = acc.reduce((prev, curr) => prev + curr);
                acc = [];
                acc.push(sum);
            }
            break;
        case "prod":
            {
                const product = acc.reduce((prev, curr) => prev * curr);
                acc = [];
                acc.push(product);
            }
            break;
        default:
            acc.push(Number.parseFloat(op));
            break;
    }
    return acc;
}

File: ScalaRewrite/src/main/scala/assignments/extra/Rpn.scala
package assignments.extra

import scala.collection.mutable

def rpn(input: String): Option[Float] =
  val numbers = mutable.Stack.from(input.split(' ')).foldLeft(mutable.Stack.empty[Float])(innerRpn)
  val result = numbers.pop()
  if !numbers.isEmpty then None else Some(result)

private def innerRpn(numbers: mutable.Stack[Float], op: String): mutable.Stack[Float] =
  op match
    case "+" | "-" | "*" | "/" | "^" =>
      val n1 = numbers.pop()
      val n2 = numbers.pop()
      op match
        case "+" => numbers.push(n2 + n1)
        case "-" => numbers.push(n2 - n1)
        case "*" => numbers.push(n2 * n1)
        case "/" => numbers.push(n2 / n1)
        case "^" => numbers.push(Math.pow(n2.toDouble, n1.toDouble).toFloat)
      numbers
    case "ln" =>
      val n1 = numbers.pop()
      numbers.push(Math.log(n1.toDouble).toFloat)
    case "log10" =>
      val n1 = numbers.pop()
      numbers.push(Math.log10(n1.toDouble).toFloat)
    case "sum" =>
      mutable.Stack(numbers.sum)
    case "prod" =>
      mutable.Stack(numbers.product)
    case _ =>
      numbers.push(op.toFloat)